src/testsuite/arith128_test_char.c

```diff
@@ -296,6 +296,308 @@ test_popcntb (void)
   return (rc);
 }
 
+//#define __DEBUG_PRINT__ 1
+int
+test_mrgeob (void)
+{
+  vui8_t i, j;
+  vui8_t k, e;
+  int rc = 0;
+
+  printf ("\ntest_mrgeob Vector Merge Even/Odd Bytes\n");
+
+  i = (vui8_t) { 10, 1, 20, 2, 30, 3, 40, 4,
+		50, 5, 60, 6, 70, 7, 80, 8 };
+  j = (vui8_t) { 90, 9, 0xa0, 10, 0xb0, 11, 0xc0, 12,
+		 0xd0, 13, 0xe0, 14, 0xf0, 15, 0xff, 0 };
+  e = (vui8_t) { 10, 90, 20, 160, 30, 0xb0, 40, 0xc0,
+		 50, 0xd0, 60, 0xe0, 70, 0xf0, 80, 0xff };
+  k = vec_mrgeb (i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint8x ("mrgeb(\t{", i);
+  print_vint8x ("\t\t{", j);
+  print_vint8x ("\t\t{", k);
+#endif
+  rc += check_vuint128x ("vec_mrgeb:", (vui128_t) k, (vui128_t) e);
+
+  e = (vui8_t)
+	  { 1, 9, 2, 10, 3, 11, 4, 12, 5, 13, 6, 14, 7, 15, 8, 0 };
+  k = vec_mrgob (i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint8x ("mrgeo(\t{", i);
+  print_vint8x ("\t\t{", j);
+  print_vint8x ("\t\t{", k);
+#endif
+  rc += check_vuint128x ("vec_mrgob:", (vui128_t) k, (vui128_t) e);
+
+  return (rc);
+}
+
+int
+test_mrgahlb (void)
+{
+  vui16_t i, j;
+  vui8_t k, e;
+  int rc = 0;
+
+  printf ("\ntest_mrgahlb Vector Merge Algebraic High/Low Bytes\n");
+
+  i = (vui16_t)CONST_VINT128_H(0xf101, 0xf202, 0xf303, 0xf404, 0xf505, 0xf606, 0xf707, 0xf808);
+  j = (vui16_t)CONST_VINT128_H(0xf909, 0xfa0a, 0xfb0b, 0xfc0c, 0xfd0d, 0xfe0e, 0xff0f, 0xf000);
+  e = (vui8_t)(vui16_t)CONST_VINT128_H(0xf1f9, 0xf2fa, 0xf3fb, 0xf4fc,
+			      0xf5fd, 0xf6fe, 0xf7ff, 0xf8f0);
+  k = vec_mrgahb(i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint16x ("mrgahb(\t{", i);
+  print_vint16x ("\t\t{", j);
+  print_vint8x  ("\t\t{", k);
+#endif
+  rc += check_vuint128x ("vec_mrgahb:", (vui128_t)k, (vui128_t) e);
+
+  e = (vui8_t)CONST_VINT128_H(0x1, 0x5, 0x2, 0x6, 0x3, 0x7, 0x4, 0x8);
+  e = (vui8_t)(vui16_t)CONST_VINT128_H(0x0109, 0x020a, 0x030b, 0x040c,
+			      0x050d, 0x060e, 0x070f, 0x0800);
+  k = vec_mrgalb(i, j);
+
+#ifdef __DEBUG_PRINT__
+  print_vint16x ("mrgalb(\t{", i);
+  print_vint16x ("\t\t{", j);
+  print_vint8x  ("\t\t{", k);
+#endif
+  rc += check_vuint128x ("vec_mrgalb:", (vui128_t)k, (vui128_t) e);
+
+  return (rc);
+}
+
+int
+test_mulhub (void)
+{
+  vui8_t i, j;
+  vui8_t k, e;
+  int rc = 0;
+
+  printf ("\ntest_mulhub Vector Multiply High Unsigned Bytes\n");
+
+  i = (vui8_t)
+	  { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 };
```

Comment:
But this is HOW the ISA works. Something you need to learn.

The splat immediate instructions only have a 5-bit signed immediate field. 
So this initializer generates one vspltisb instruction. A latency of 2 cycles.
Using 255 would require a load and 3-4 instructions. A latency of 9 cycles, best case.
Similar for halfword and word.
POWER9 add xxspltib with an 8-bit immediate field but only fpr splat byte (not halfword or word) and I have not seen the compiler smart enough to use it.

I have no plans to change this.

Reply:
Yep, learning as I go.  So this works because the compiler uses "Vector Splat Immediate Signed Byte" to initialize a vector of _unsigned_ bytes?  So you're actually using -1 to get 255, _and_ using 255 would be suboptimal, even though you get the same result?  The average programmer (like me?) is not going to get that.  I wonder if a comment is warranted.  Seems like the compiler could figure that out, too, such that 255 and any other 5-bit equivalents after sign extension would use vspltisb.  I certainly prefer more readable code and a smarter compiler.

Reply:
How about:

  // Same as vec_splats ((unsigned char)255), only faster
  i = (vui8_t) vec_splat_u8 (-1);

Reply:
FWIW, using 240-255 will indeed be emitted as vspltisb:
```
$ cat splat.c
#include <altivec.h>

vector unsigned char f() {
        return (vector unsigned char) { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 };
}

vector unsigned char g() {
        return (vector unsigned char) { 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255 };
}

vector unsigned char h() {
        return (vector unsigned char) { 240, 240, 240, 240, 240, 240, 240, 240, 240, 240, 240, 240, 240, 240, 240, 240 };
}
$ gcc -c -O3 splat.c
$ objdump --disassemble splat.o
splat.o:     file format elf64-powerpc
0000000000000000 <.f>:
   0:   10 5f 03 8c     vspltisw v2,-1
   4:   4e 80 00 20     blr
        ...
  14:   60 00 00 00     nop
  18:   60 00 00 00     nop
  1c:   60 42 00 00     ori     r2,r2,0

0000000000000020 <.g>:
  20:   10 5f 03 8c     vspltisw v2,-1
  24:   4e 80 00 20     blr
        ...
  34:   60 00 00 00     nop
  38:   60 00 00 00     nop
  3c:   60 42 00 00     ori     r2,r2,0

0000000000000040 <.h>:
  40:   10 50 03 0c     vspltisb v2,-16
  44:   4e 80 00 20     blr
        ...
```